File: backend/app/routers/details.py
from typing import Annotated

from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.database.session import get_db_session
from app.schemas.detail import (
    DetailCreate,
    DetailPartialUpdate,
    DetailReadFull,
    DetailUpdate,
)
from app.services.detail_service import detail_service


router = APIRouter(prefix="/details", tags=["details"])

SessionDep = Annotated[AsyncSession, Depends(get_db_session)]


@router.get("", response_model=list[DetailReadFull])
async def get_details(
    session: SessionDep,
    skip: Annotated[int, Query(ge=0)] = 0,
    limit: Annotated[int, Query(ge=1, le=100)] = 100,
):
    return await detail_service.get_details(session, skip=skip, limit=limit)


@router.get("/{detail_id}", response_model=DetailReadFull)
async def get_detail(detail_id: int, session: SessionDep):
    return await detail_service.get_detail_by_id(session, detail_id)


@router.post(
    "",
    response_model=DetailReadFull,
    status_code=status.HTTP_201_CREATED,
)
async def create_detail(payload: DetailCreate, session: SessionDep):
    return await detail_service.create_detail(session, payload)


@router.put("/{detail_id}", response_model=DetailReadFull)
async def update_detail(
    detail_id: int,
    payload: DetailUpdate,
    session: SessionDep,
):
    return await detail_service.update_detail(session, detail_id, payload)


@router.patch("/{detail_id}", response_model=DetailReadFull)
async def partial_update_detail(
    detail_id: int,
    payload: DetailPartialUpdate,
    session: SessionDep,
):
    return await detail_service.partial_update_detail(session, detail_id, payload)


@router.delete("/{detail_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_detail(detail_id: int, session: SessionDep) -> None:
    await detail_service.delete_detail(session, detail_id)

File: frontend/src/api/details.ts
import { apiClient } from "./client";
import type {
  Detail,
  DetailCreate,
  DetailFilters,
  DetailPatch,
  DetailQuantityUpdate,
  DetailUpdate,
} from "../types/detail";

export async function getDetails(params?: DetailFilters): Promise<Detail[]> {
  const response = await apiClient.get<Detail[]>("/details", { params });
  return response.data;
}

export async function getDetail(id: number): Promise<Detail> {
  const response = await apiClient.get<Detail>(`/details/${id}`);
  return response.data;
}

export async function createDetail(data: DetailCreate): Promise<Detail> {
  const response = await apiClient.post<Detail>("/details", data);
  return response.data;
}

export async function updateDetail(
  id: number,
  data: DetailUpdate,
): Promise<Detail> {
  const response = await apiClient.put<Detail>(`/details/${id}`, data);
  return response.data;
}

export async function patchDetail(
  id: number,
  data: DetailPatch,
): Promise<Detail> {
  const response = await apiClient.patch<Detail>(`/details/${id}`, data);
  return response.data;
}

export async function deleteDetail(id: number): Promise<void> {
  await apiClient.delete(`/details/${id}`);
}

export async function getLowStockDetails(threshold = 5): Promise<Detail[]> {
  const response = await apiClient.get<Detail[]>("/details/low-stock", {
    params: { threshold },
  });
  return response.data;
}

export async function updateDetailQuantity(
  id: number,
  quantity: number,
): Promise<Detail> {
  const payload: DetailQuantityUpdate = { quantity };
  const response = await apiClient.patch<Detail>(
    `/details/${id}/quantity`,
    payload,
  );
  return response.data;
}
